### ISSUE TITLE
Slow GaudiExec submission

### ISSUE THREAD

**Issue by olupton** [ganga-devs/ganga#929]
Hi,

I'm submitting several jobs that all use the same local `DaVinciDev_v42r1/` area.
It seems that every job builds, copies, uploads and removes the `cmake-input-sandbox.tgz` file.
This seems to be rather slow, and unnecessary (I don't think this file contains the job options?).
Couldn't `make` be trusted to update this archive if needed?

Thanks,
Olli

**Comment by egede** [comment]
@olupton So do I understand correctly that you are forced to ```unprepare``` the application object because you want to change the list of options files used in the application object? But everything else is the same?

**Comment by olupton** [comment]
Hi @egede,
At the moment I am not calling `unprepare()`, I'm making several `GaudiExec` instances that all use the same `DaVinciDev` area (but have different `options`) and calling `prepare()` on each.
I could try changing to
```
app = GaudiExec()
job1.application = GaudiExec()
job1.application.options = [ ... ]
job1.application.prepare()
job2.application = job1.application
job2.application.unprepare()
job2.application.options = [ ... ]
job2.application.prepare()
```
but my guess was that this would be essentially the same?

**Comment by rob-c** [comment]
@olupton Would it be worth simply creating a single prepared application and copy it into multiple jobs?
(I think there is an (LHCb specific?) setting about whether we un-prepare automatically on copy. Either way if you make sure you have `[Preparable]unprepare_on_copy=False` in your `.gangarc` it should work)

**Comment by olupton** [comment]
@rob-c I was under the impression that a prepared application includes the options files?
e.g. I see
```
$ ls /workspace/gangadir/shared/olupton/conf-f6351380-8725-4a18-91bd-440284d59d1d/
cmake-input-sandbox.tgz  DaVinci-DataType-2012.py  DaVinci-EvtMax--1.py  DaVinci-Simulation-False.py  new_buggy.py  tuple_script.py
```
I am already making sure that the same prepared application is used for jobs that have identical options in addition to an identical build area.

**Comment by rob-c** [comment]
@olupton Hmm, yes I see your problem now. I'll give some thougt to this as I did want to have multiple prepared files for the GaudiExec class to avoid this but there were some concerns about serialization of the XML raised which got in the way of doing this which were never resolved.
I'll see if I can do something clever in a short PR.

**Comment by rob-c** [comment]
@olupton What is the main _slowness_ in the job submission? is it the preparation/build or the upload steps?

**Comment by egede** [comment]
@olupton I think what we talk about here is the difference between the packing up of the release area (what ```prepare``` does) and the specification of the job (which options to give to the gaudirun.py) script. It would not harm the jobs above if ```job1.py``` and ```job2.py``` were in the same shared area as long as I could request that job 1 is using the first options file and job2 the second.

```
j=Job()
myApp1 = GaudiExec()
myApp1.directory = "$SOMEPATH/DaVinciDev_v40r2"
myApp1.options = ["$SOMEPATH/DaVinciDev_v40r2/myDaVinciOpts_1.py"]
myApp1.prepare()

myApp2 = myApp1.copy()

# You would like to be able to do this without losing the prepared area
# through an unprepare call.
myApp1.options = ["$SOMEPATH/DaVinciDev_v40r2/myDaVinciOpts_2.py"]
```

I think this could be achieved through exploring the difference that we already have between ```ProtectedAttributeError``` and ```ReadOnlyObjectError```. So in this case ```myApp1.directory``` would throw an ```ProtectedAttributeError``` after the prepare state, but ```myApp1.options``` would be allowed to be changed. After submission, both would throw ```ReadOnlyObjectError```.

**Comment by olupton** [comment]
@rob-c, my script that calls `prepare()` but not `submit()` is very slow. Calling `submit()` afterwards isn't so painful, but the output:
```
INFO     Uploading file '/tmp/olupton/__conf-ce340885-f69f-4cae-8600-f82790853a89.tgz' to 'IN2P3-USER' as '/lhcb/user/o/olupton/GangaInputFile/Job_1976/__conf-ce340885-f69f-4cae-8600-f82790853a89.tgz'
INFO     Uploading file '/workspace/gangadir/workspace/olupton/LocalXML/1976/input/jobScripts-27b8300c-5d68-4bfa-b331-d7b8c74bc278.tar.gz' to 'PIC-USER' as '/lhcb/user/o/olupton/GangaInputFile/Job_1976/jobScripts-27b8300c-5d68-4bfa-b331-d7b8c74bc278.tar.gz'
```
suggests that my local software changes are going to be uploaded multiple times if I submit several jobs that share a prepared application.

@egede so you mean my various `optionsN.py` files would be bundled up inside `cmake-input-sandbox.tgz`? This could work, although it seems a little hacky and would require you to know all the jobs' options before you prepared any of them. Looking at the output above, it might be neater if the options files were shipped along with the worker node scripts in `jobScripts-XXX.tar.gz` instead of `__conf-XXX.tgz`? If the job number was removed from the `conf-XXX` grid path then you could also avoid uploading this multiple times just be checking if the file already exists on the grid?

**Comment by rob-c** [comment]
@olupton yes, for now the DIRAC upload is done in 1 (potentially large) file containing both the cmake and options files (and inputfiles) rather than 2 files (1 for software and 1 for everything else which may change). Given the high speeds accessible on lxplus I imagine the main slowdown is in DIRAC doing what needs to be done before it will upload a file.
The files could be renamed but they're not useful when moving between files as there is no way for Ganga to tell that the directory on disk wasn't touched since it submitted the last job.
I can add a step to allow you to upload the same cmake target for multiple jobs but rebuilding this is a minor part of the length of time to submit jobs I suspect...

**Comment by olupton** [comment]
@rob-c, sure, no point making changes that won't make much difference. Although the output suggests that there are already 2 files uploaded for every job (software+options, and the ganga-generated job scripts).
In my current setup the main delay is certainly re-making the `cmake-input-sandbox.tgz` for every software+options combination.

**Comment by rob-c** [comment]
OK, I can add a trivial option to the schema to allow for passing pre-built targets if the target building takes far too long and should be reproducible.

**Comment by egede** [comment]
@olupton Yes, I agree that the options could be put separately. There are a few issues with this though. If the options file *is* inside the release area, it will be packed up twice and it should be well defined which version is used (the one from the options attribute). Also if file A is including file B, then a possible new version of file B will only be included if it is given in the options attribute (i.e. at submit time we do not recursively search the includes).

**Comment by egede** [comment]
@rob-c I am not sure that passing pre-built targets is the bedst way to solve this problem.

**Comment by olupton** [comment]
@egede not sure I got all the details there. If I put a bonus `test.py` in `DaVinciDev_vXrY/` then it doesn't seem to end up in `cmake-input-sandbox.tgz`, so your example code
```
j=Job()
myApp1 = GaudiExec()
myApp1.directory = "$SOMEPATH/DaVinciDev_v40r2"
myApp1.options = ["$SOMEPATH/DaVinciDev_v40r2/myDaVinciOpts_1.py"]
myApp1.prepare()
```
would I think at the moment result in a `conf-XXX.tgz` containing `myDaVinciOpts_1.py` and `cmake-input-sandbox.tgz` at the top level, and without `myDaVinciOpts_1.py` present in the `cmake-input-sandbox.tgz`. This is what I already have, isn't it?

I think I can workaround with `extraArgs`:
```
j = Job()
j.application = GaudiExec()
j.application.options = [ '/path/to/dummy.py' ] # can't leave it empty...otherwise I get an error on submit
j.application._impl.extraArgs = [ 'options1.py' ] # _impl because of a bug (I think), see below
j.inputfiles = [ '/path/to/options1.py' ]
j.application.prepare()
j2 = j.copy()
j2.application._impl.extraArgs = [ 'options2.py' ]
j2.inputfiles = [ '/path/to/options2.py' ]
```
The `_impl` is because `extraArgs` seems confused about what you can assign to it.
`j2.application.extraArgs=[['foo']]` works, but ends up producing an error like
```
WARNING  ---------- error in user/extension code ----------
WARNING  Traceback (most recent call last):
WARNING    File "/cvmfs/lhcb.cern.ch/lib/lhcb/GANGA/GANGA_v603r1/install/ganga/python/Ganga/GPIDev/Lib/Job/Job.py", line 1498, in submit
WARNING      jobmasterconfig = self._getJobMasterConfig()
WARNING    File "/cvmfs/lhcb.cern.ch/lib/lhcb/GANGA/GANGA_v603r1/install/ganga/python/Ganga/GPIDev/Lib/Job/Job.py", line 1196, in _getJobMasterConfig
WARNING      jobmasterconfig = rtHandler.master_prepare(self.application, appmasterconfig)
WARNING    File "/cvmfs/lhcb.cern.ch/lib/lhcb/GANGA/GANGA_v603r1/install/ganga/python/GangaLHCb/Lib/RTHandlers/GaudiExecRTHandlers.py", line 387, in master_prepare
WARNING      generateDiracScripts(app)
WARNING    File "/cvmfs/lhcb.cern.ch/lib/lhcb/GANGA/GANGA_v603r1/install/ganga/python/GangaLHCb/Lib/RTHandlers/GaudiExecRTHandlers.py", line 312, in generateDiracScripts
WARNING      generateJobScripts(app, appendJobScripts=True)
WARNING    File "/cvmfs/lhcb.cern.ch/lib/lhcb/GANGA/GANGA_v603r1/install/ganga/python/GangaLHCb/Lib/RTHandlers/GaudiExecRTHandlers.py", line 297, in generateJobScripts
WARNING      wnScript = generateWNScript(prepareCommand(this_app), this_app)
WARNING    File "/cvmfs/lhcb.cern.ch/lib/lhcb/GANGA/GANGA_v603r1/install/ganga/python/GangaLHCb/Lib/RTHandlers/GaudiExecRTHandlers.py", line 153, in prepareCommand
WARNING      full_cmd += " " + " ".join(app.extraArgs)
WARNING  TypeError: sequence item 0: expected string, list found
WARNING  --------------------------------------------------
ERROR    sequence item 0: expected string, list found ... reverting job 2112 to the new status
INFO     job 2112 status changed to "new"
ERROR    Error: JobError: Error: sequence item 0: expected string, list found
```
so it seems a list of strings is expected (which makes sense), but you're not allowed to assign one...

I'll try testing this in my actual use-case...

Incidentally, after the move to cmake have I understood correctly that we no longer pickle options? The options files are always parsed directly on every worker node?

**Comment by egede** [comment]
@olupton Thanks for the further input. Again need to think about best implementation to achieve what seems a very reasonable optimisation.

@olupton You are correct that options are no longer pickled. With the old ```Gaudi()``` type applications there was no way around it, but I always regarded it as fragile.

**Comment by olupton** [comment]
Hi,

My workaround above seems to avoid the main problem. Job creation is now limited by how slow
```
diracfiles = [ DiracFile(lfn=str(lfn)) for lfn in lfn_list ]
dataset = LHCbDataset(files = diracfiles)
``` 
is...which is another matter (though if there's an easy way of making this faster I'd like to hear it...I have no idea why it's so slow).

One annoyance is that now I have multiple jobs sharing the same conf-XXX directory then if I submit them in parallel they tread on each other by all using the same `/tmp/olupton/__conf-XXX.tgz` path. I guess it would be trivial to put the job ID into the filename and avoid this problem.

Another thing is that the different jobs' `inputfiles` end up being mixed in together, i.e. in the example above then the `__conf-XXX.tgz` that gets uploaded to the grid contains both `options1.py` and `options2.py` despite neither job requesting both. After submission these files are also both present in `/workspace/gangadir/shared/olupton/conf-XXX`. This isn't actually a problem in my case, but it's a bit strange.

**Comment by egede** [comment]
Hi Olli,

Can I get you to file a separate bug on the race condition where the same 
name is used in /tmp.

Ulrik

**Comment by rob-c** [comment]
@olupton The problem with creating a large list of files from LFN is unfortunately very well known and one I've spent significant time looking into.
There is a bug somewhere in the time it takes to allocate memory for ~10k DiracFile classes and despite replacing most of the chain with classes which use `__slots__` I can't make this particular problem go away entirely.

There is PR #917 where I have tried to address this further and have had some successes but this will need review. All the branch does is cache some large(ish) list lookups as well as make more extensive use of `__slots__` which seems to improve the ganga shutdown experience. The mechanism for allowing for flexible editting of the class definitions at runtime does step in the way of optimising class construction from a technical perspective. We can show this as creating 10k dummy classes takes a modest amount of time compared to 10k DiracFile objects the problem lies somewhere in Ganga for this.
There is the risk with this PR if you edit the ganga config at runtime that the changes aren't taken into account but this is mitigated by making sure the runtime is defined in the .gangarc before launching.

I was assured this wouldn't happen with uuids and I was originally using the job name as a unique element for each job. Are you copying the application between jobs as this could be (incorrectly) copying the uuid across which could explain this, otherwise I'm concerned that this is possible. (I think prepared state jobs are being copied and if we support caching the prepared state there is some pollution of the prepared state dir vs the local dir, but this is required to allow the job to run when changing back to a local backend for debugging.)
This _should_ only require a trivial PR to fix.

**Comment by alexanderrichards** [comment]
@rob-c I havent fully read the problem that this issue is talking about but something just triggered in the back of my brain. Is the slowdown due to in initialisation of each of the ~10K DiracFile classes or is it due to the ever expanding list appending that I assume is happening?

The reason I ask is that I've noted that some people have seen slight speed increases with a sort of pseudo-preallocation of the list as below:

```python
mylist = [None] * 10000
for i in xrange(10000):
    mylist[i] = DiracFile(blah)
```

Now when you scale this small speed up to 10K it may still not make a difference or it might. Just thought I'd point it out as I recall seeing it some time ago.

**Comment by olupton** [comment]
@rob-c re: the uuids, this is explicitly for jobs that *do* share prepared applications and therefore *do* have the same uuid. This is 'intentional' in that it seems to be the only way of avoiding the original problem discussed here, namely that making the release area tarball is very slow.
The problem seems to be that the prepared application uuid is incorrectly assumed to be unique for the job as well.

I opened https://github.com/ganga-devs/ganga/issues/932 as requested by @egede 

@alexanderrichards I tried that pseudo-preallocation but it didn't make a dramatic difference.

**Comment by rob-c** [comment]
@alexanderrichards OK, this may speed up some part of the code I use in testing this. The problem is though that there is a slowdown in allocating a DiracFile vs a dummy class object. A _lot_ of this I have traced down to the `getDefault` method in the `Schema` (or whatever it's actually called). I've overly optimised it in #917 and it seems to make a strong impact in these cases where it's called some ~50k times but it's still a slow piece of code as it performs about 4 or 5 list lookups every single time it's called...

**Comment by egede** [comment]
@olupton Following the suggestion from @alexanderrichards, I tried

```
Ganga In [20]: %time l = [DiracFile()] * 10000
CPU times: user 4 ms, sys: 0 ns, total: 4 ms
Wall time: 3.11 ms

Ganga In [21]: %time for f in l: f.lfn = '/lhcb/MC/MC10/ALLSTREAMS.DST/00009117/0000/00009117_00000686_1.allstreams.dst'
CPU times: user 1.89 s, sys: 228 ms, total: 2.12 s
Wall time: 1.92 s

[12:46:45]
Ganga In [22]: %time s = LHCbDataset(files=l)
CPU times: user 16.4 s, sys: 934 ms, total: 17.3 s
Wall time: 16.2 s
```

from which you can see that assigning the LFNs is fast (<2 s) but then creating the ```LHCbDataset``` object is quite slow (16 s), but I have a feeling that you observe something much slower. Actually, just replicating what you do seems reasonably fast as well

```
Ganga In [30]: %time l2 = [ DiracFile(lfn='/lhcb/MC/MC10/ALLSTREAMS.DST/00009117/0000/00009117_00000686_1.allstreams.dst') for i in xrange(10000) ]
CPU times: user 18.4 s, sys: 1.06 s, total: 19.5 s
Wall time: 18.3 s
```

This time should be compared to the time it takes to submit hundreds of subjobs, so I fail a bit to see the problem.

**Comment by rob-c** [comment]
@egede Yes but this is then an _additional_ 20sec every time the job has to be read from disk (once per session ideally but this translates into ~1min for loading a large job with I/O on AFS taking ~5sec.
We've reduced many of the un-needed duplications of objects behind the scenes but a 20-30sec cost for duplicating an object already in memory leads to slowdowns. I've seen users in Edinburgh complain because they've 5 or 6 jobs with a few hundred subjobs all trying to be monitored at the same time and most of the slowdown is in handling these large lists best I can tell.

**Comment by olupton** [comment]
@egede that's an unfair estimate of the list creation time, because you are only creating one `DiracFile()` object and storing 10000 references to it:
```
Ganga In [119]: %time l = [DiracFile()] * 10000
CPU times: user 5 ms, sys: 0 ns, total: 5 ms
Wall time: 3.5 ms

Ganga In [122]: %time for n, f in enumerate(l): f.lfn = '/lhcb/MC/MC10/ALLSTREAMS.DST/' + str(n)
l[-1]CPU times: user 3.98 s, sys: 1.17 s, total: 5.15 s
Wall time: 3.89 s

Ganga In [125]: l[-1].lfn
Ganga Out [125]: '/lhcb/MC/MC10/ALLSTREAMS.DST/9999'

Ganga In [126]: l[-2].lfn
Ganga Out [126]: '/lhcb/MC/MC10/ALLSTREAMS.DST/9999'
```

In my real-world use-case then as you say, the LHCbDataset construction is about the same speed as making the list of DiracFile objects.

Anyway, as you say, now I worked around the build area preparation speed it's not so bad, and faster than the job submission. It's just a bit annoying because this is something I want to do ~interactively, while the submission is "fire-and-forget"...

**Comment by egede** [comment]
@olupton Yes, you are of course correct. So the total time I see is indeed about 40s. There is a bit annoying thing here in that when you create the ```LHCbDataset``` object, you make a deepcopy of the list. So in both cases, we just see the time for creating 10k empty ```DiracFile``` objects.

**Comment by rob-c** [comment]
@egede I've done some further profiling atop #917 and it's possible after some googling that we're hitting some problems due to the use of RLocks and the GIL. I've seen a lot of writing about RLocks being extremely cheap after Python 3.3 due to a large re-write but I've no idea if/when this could be back-ported.
I'm tempted to look into some light-weight extension of the RLock to check for the thread-ID being different and then getting the lock again but this is potentially no better than what is already in place behind the scenes.
I could be wrong but I'm seeing a large (20%+) of the amount of profile time being spent in getting the lock to execute the code that we're interested in. I'm keen to try running some single thread tests with the locking code removed to confirm 100% if this is the case or not.

**Comment by egede** [comment]
@rob-c I doubt you will save anything by rolling your own lock code. A better optimisation would be to consider how to avoid the many useless deep copy operations that we make.

**Comment by rob-c** [comment]
From my development branch I am now getting:

```
[13:10:25]
Ganga In [6]: %time lf=[LocalFile() for _ in range(10000)]
CPU times: user 1.12 s, sys: 268 ms, total: 1.39 s
Wall time: 1.15 s

[13:10:27]
Ganga In [7]: %time msf = [MassStorageFile() for _ in range(10000)]
CPU times: user 7.01 s, sys: 2.05 s, total: 9.06 s
Wall time: 6.95 s

[13:10:36]
Ganga In [8]: %time df=[DiracFile() for _ in range(10000)]
CPU times: user 7.2 s, sys: 1.27 s, total: 8.47 s
Wall time: 7.1 s
```

Which is a large improvement in the speed of constructing objects. I think this needs further testing but I'll try and land the branch later this week / next.

**Comment by rob-c** [comment]
I think this will be fixed largely by #917 hopefully in Ganga 6.7

**Comment by egede** [comment]
Many changes implemented in this area.
